test/hotspot/jtreg/serviceability/jvmti/vthread/StopThreadTest/StopThreadTest.java

```diff
@@ -0,0 +1,242 @@
+    static native void resumeThread(Thread thread);
+    static native void ensureAtBreakpoint();
+    static native void notifyAtBreakpoint();
+    static native int  stopThread(Thread thread);
+
+    static int status = PASSED;
+
+    static void setFailed(String msg) {
+        log("\nFAILED: " + msg);
+        status = FAILED;
+    }
+
+    static void throwFailed(String msg) {
+        log("\nFAILED: " + msg);
+        throw new RuntimeException("StopThreadTest failed!");
+    }
+
+    public static void main(String args[]) {
+        run();
+        if (status == FAILED) {
+            throwFailed("StopThreadTest!");
+        }
+        log("\nStopThreadTest passed");
+    }
+
+    public static void run() {
+        TestTask testTask = new TestTask();
+        Thread testTaskThread = null;
+        AssertionError excObject = new AssertionError();
+        int retCode;
+
+        prepareAgent(TestTask.class, excObject);
+
+        log("\nMain #A: method A() must be blocked on entering a synchronized statement");
+        synchronized (TestTask.lock) {
+            testTaskThread = Thread.ofVirtual().name("TestTaskThread").start(testTask);
+            testTask.ensureStarted();
+
+            log("\nMain #A.1: unsuspended");
+            retCode = stopThread(testTaskThread);
+            if (retCode != THREAD_NOT_SUSPENDED) {
+                throwFailed("Main #A.1: expected THREAD_NOT_SUSPENDED instead of: " + retCode);
+            } else {
+                log("Main #A.1: got expected THREAD_NOT_SUSPENDED");
+            }
+
+            log("\nMain #A.2: suspended");
+            suspendThread(testTaskThread);
+            retCode = stopThread(testTaskThread);
+            if (retCode != JVMTI_ERROR_NONE) {
+                throwFailed("Main #A.2: expected JVMTI_ERROR_NONE instead of: " + retCode);
+            } else {
+                log("Main #A.2: got expected JVMTI_ERROR_NONE");
+            }
+            resumeThread(testTaskThread);
+        }
+        log("\nMain #B: method B() must be blocked in a breakpoint event handler");
+        {
+            ensureAtBreakpoint();
+
+            log("\nMain #B.1: unsuspended");
+            retCode = stopThread(testTaskThread);
+            if (retCode != THREAD_NOT_SUSPENDED) {
+                throwFailed("Main #B.1: expected THREAD_NOT_SUSPENDED instead of: " + retCode);
+            }
+
+            log("\nMain #B.2: suspended");
+            suspendThread(testTaskThread);
+            retCode = stopThread(testTaskThread);
+            if (retCode != JVMTI_ERROR_NONE) {
+                throwFailed("Main #B.2: expected JVMTI_ERROR_NONE");
+            }
+            resumeThread(testTaskThread);
+
+            notifyAtBreakpoint();
+        }
+
+        log("\nMain #C: method C() sends AssertionError object to its own thread");
+        {
+            // StopThread is expected to succeed.
+            testTask.ensureFinished();
+        }
```

Comment:
I don't see how this is doing any testing. Where is the stopThread(null) call?

Reply:
The target virtual thread just continues and invokes method `C()` which sends asynchronous exception with JVMTI `StopThread` to current thread:
```
        // This method uses StopThread to send an AssertionError object to
        // its own thread. It is expected to succeed.
        static void C() {
            log("TestTask.C: started");
            StopThreadTest.stopThread(Thread.currentThread());
            log("TestTask.C: finished");
        }
```
I'll try to add a comment to make it clear.